### PR TITLE
chore: Make main branch buildable

### LINKED ISSUE #505: Error parsing the Package.swift file
## Description

Hello! I'm part of the [Swift Package Index](https://swiftpackageindex.com) team and was just trying to add this package to the index. We use `swift package dump-package` to extract package metadata in as JSON but when running that command for this package, the following error occurs:

```
$ xcrun swift package dump-package
/Users/dave/Desktop/aws-sdk-swift: error: manifest parse error(s):
main/Package.swift:104: Fatal error: 'try!' expression unexpectedly raised an error: Error Domain=NSCocoaErrorDomain Code=260 "The folder “release” doesn’t exist." UserInfo={NSFilePath=/Users/dave/Projects/Amplify/SwiftSDK/aws-sdk-swift/release, NSUserStringVariant=(
    Folder
), NSUnderlyingError=0x600000a28090 {Error Domain=NSOSStatusErrorDomain Code=-43 "fnfErr: File not found"}}
error: invalid manifests at [<AbsolutePath:"/Users/dave/Desktop/aws-sdk-swift">]
```

This is broken on `main` but works on the `0.1.0` tag. Unfortunately, to add the package to the index we need to be able to initially run the `dump-package` on a checkout of the default branch.

## Environment

* AWS Swift SDK version used: This is broken on `main` but works on the `0.1.0` tag.
* Operating System and version (iOS/macOS/Linux): macOS Monterey
* Xcode version (if applicable): Version 13.1 (13A1030d)
* Swift compiler version (if applicable): Apple Swift version 5.5.1 (swiftlang-1300.0.31.4 clang-1300.0.29.6)
* Swift Tools version (if applicable): Apple Swift version 5.5.1 (swiftlang-1300.0.31.4 clang-1300.0.29.6)

## Reproduction Steps

* Check out the `main` branch of this repository locally.
* Change into the checked out directory.
* Run `xcrun swift package dump-package`

## Observed Behavior

The manifest fails to parse.

## Expected Behavior

The manifest parses successfully.

---

Thank you!

### LINKED ISSUE #620: Unpublished package does not work without extra setup
## Description
The AWS Swift SDK does not work when installed from the `main` branch unless significant local setup is performed (see the [steps](https://github.com/awslabs/aws-sdk-swift/blob/main/Package.swift#L20) listed in this repo's `Package.swift` file.)

It is desired that a tester could require this package like so:
```
.package(url: "https://github.com/awslabs/aws-sdk-swift.git", branch: "main")
```
and have it install from this repo's `main` branch, with the Swift SDK already generated, and dependencies `smithy-swift` and `aws-crt-swift` installed from their `main` branches, too.

## Context
This is causing difficulty for testers who are not Swift SDK developers, but are attempting to use & test unreleased changes to the SDK.

### PR DESCRIPTION
**Pull request by jbelkins**
## Issue \#
Fixes #505
Fixes #620 

### Summary
The "release" `Package.swift` may now be merged back to the `main` branch following a release, because the special logic in the current `Package.swift` that is used to adapt the manifest for development and CI use is now provided by the manifest generator script.

### Details
The package manifest generation script (`scripts/generatePackageSwift.swift`) has been improved to allow for generation of   a manifest that is free of hard-coded paths and special logic, and that may depend upon `smithy-swift` or `aws-crt-swift` referred to by published version, branch or local file path.

The contents of `Package.swift` are now determined entirely by these inputs:
- The generation script `scripts/generatePackageSwift.swift`
- The contents of `release/` determine which AWS clients are published.
- The settings in `versionDependencies.plist` determine the version, branch, or path where dependencies are located.
- The env var `AWS_SDK_RELEASE_IN_PROGRESS` (when set) generates the manifest for a release (dependencies lock to published versions only.)
- Env vars `AWS_SDK_SWIFT_CI_DIR`, `AWS_CRT_SWIFT_CI_DIR`, and `SMITHY_SWIFT_CI_DIR` (when set) override `versionDependencies.plist` and set local paths to dependencies.  (This is for compatibility with the AWS CRT builder script used on our CI tasks.)

No other factors affect the manifest contents, and the manifest may be regenerated any time one of these factors changes.  The `Package.swift` file should no longer be edited by hand, since edits may be overwritten by the script at any time.

(Note that the `codegen/Package.swift` manifest used for protocol tests is unchanged, and protocol tests still require that the project be installed at `~/Projects/Amplify/SwiftSDK` to run on local.)

When developing on the desktop, developers may use the "package edit mode" features of Xcode or Swift Package Manager to develop against local copies of dependencies rather than edit the `Package.swift` file.

This will allow the release manifest to be merged back into `main` on each release, eliminating the current `Package.swift` custom logic which presupposes the local file system location of the SDK and its dependencies.  Once a release is merged back in, a SDK consumer will be able to use the SDK from `main` without additional steps to set it up.  (As noted above, this will resolve #505 & #620.)

This PR has been verified to build successfully when built by another project as a CRT builder script downstream dependency (see https://github.com/awslabs/smithy-swift/pull/456 which builds this branch as a downstream.)

This PR has also been verified to build a release successfully ( see #667).

Specific changes:
- `scripts/generatePackageSwift.swift` is improved and derived from #650 with support for:
  - Published versions, branch names or local file paths specified in `versionDependencies.plist`
  - Local file paths in env vars `AWS_SDK_SWIFT_CI_DIR`, `AWS_CRT_SWIFT_CI_DIR`, and `SMITHY_SWIFT_CI_DIR`
  - Ignores branch name and local file path settings when building with env var `AWS_SDK_RELEASE_IN_PROGRESS` set
- Adds a pre-build step to the CRT builder, which dynamically generates `Package.swift` with file paths pointing to locally installed dependencies when running on CI
- During release, specify that the package manifest should be generated for release (i.e. with exact versions specified) by setting `AWS_SDK_RELEASE_IN_PROGRESS`

## New/existing dependencies impact assessment, if applicable
No new dependencies were added to this change.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -118,25 +19,29 @@ let package = Package(
         .iOS(.v13)
     ],
     products: [
-        .library(name: "AWSClientRuntime", targets: ["AWSClientRuntime"])
+        .library(name: "AWSClientRuntime", targets: ["AWSClientRuntime"]),
+    ],
+    dependencies: [
+        .package(url: "https://github.com/awslabs/smithy-swift", .exact("0.3.0")),
+        .package(url: "https://github.com/awslabs/aws-crt-swift", .exact("0.2.2"))
```
Comment (epau) [inline Package.swift:26]: We are targeting newer versions now
```suggestion
        .package(url: "https://github.com/awslabs/smithy-swift", .exact("0.4.0")),
        .package(url: "https://github.com/awslabs/aws-crt-swift", .exact("0.3.0"))
```

File: scripts/generatePackageSwift.swift [hunk 2]
```diff
@@ -43,57 +94,76 @@ let package = Package(
...
 
 func generateDependencies(versions: VersionDeps) {
+    let crtSwiftDependency = dependency(url: "https://github.com/awslabs/aws-crt-swift",
+                                        version: versions.awsCRTSwiftVersion,
+                                        branch: versions.awsCRTSwiftBranch, 
+                                        path: versions.awsCRTSwiftPath)
+    let clientRuntimeDependency = dependency(url: "https://github.com/awslabs/smithy-swift",
+                                             version: versions.clientRuntimeVersion,
+                                             branch: versions.clientRuntimeBranch,
+                                             path: versions.clientRuntimePath)
```
Comment (epau) [inline scripts/generatePackageSwift.swift:113]: Nit: formatting
```suggestion
    let crtSwiftDependency = dependency(
        url: "https://github.com/awslabs/aws-crt-swift",
        version: versions.awsCRTSwiftVersion,
        branch: versions.awsCRTSwiftBranch, 
        path: versions.awsCRTSwiftPath
    )
    let clientRuntimeDependency = dependency(
        url: "https://github.com/awslabs/smithy-swift",
        version: versions.clientRuntimeVersion,
        branch: versions.clientRuntimeBranch,
        path: versions.clientRuntimePath
    )
```

File: scripts/setup_for_crt_builder.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+set -x
+
+# Generates the Package.swift file, configured to work properly when using the
+# CRT builder script.
+# For use only as a pre-build step when using the CRT builder script on CI.
```
Comment (epau) [inline scripts/setup_for_crt_builder.sh:7]: What does this script have to do with CRT or the CRT builder script? From this PR it appears we're using this script as part of the aws-sdk-swift builder script.

Comment (jbelkins) [reply scripts/setup_for_crt_builder.sh:7]: When we run CI tasks, you'll see that the 1st thing we do is retrieve and install a script from a Cloudfront distro.  That is the "CRT builder script" I refer to.  (It's on Github [here](https://github.com/awslabs/aws-crt-builder).)

That script installs `aws-sdk-swift`, `smithy-swift`, and `aws-crt-swift` to the local filesystem.  Env vars in the CI job point to the specific locations on the filesystem.  

This script rewrites the `Package.swift` locally to require `smithy-swift` and `aws-crt-swift` at the locations in those env vars.

File: Package.swift (hunk 1 above)
Comment (jbelkins) [reply Package.swift]: Pinned versions have been caught up to the latest release.

File: scripts/setup_for_crt_builder.sh (hunk 3 above)
Comment (epau) [reply scripts/setup_for_crt_builder.sh:7]: ahhh TIL thank you. 
it might be helpful to include the link to that repo in the comments here

**Review by epau — APPROVED** [review]

File: scripts/setup_for_crt_builder.sh (hunk 3 above)
Comment (jbelkins) [reply scripts/setup_for_crt_builder.sh:7]: I will add the link to the build script repo to the comments in this script.

To be clear, we don't _need_ the CRT builder script to build any of our projects on CI, but most of our CI tasks are currently based on it (not the lint tasks).  Ripping it out will be a project for another day.

**Review by ganeshnj — APPROVED** [review]

File: versionDependencies.plist [hunk 4]
```diff
@@ -6,5 +6,9 @@
 	<string>0.5.0</string>
 	<key>awsCRTSwiftVersion</key>
 	<string>0.3.1</string>
+	<key>clientRuntimeBranch</key>
+	<string>main</string>
+	<key>awsCRTSwiftBranch</key>
+	<string>main</string>
```
[the remark this answers is not in the document]
Comment (jbelkins) [reply versionDependencies.plist:12]: Name changed.

File: scripts/generatePackageSwift.swift [hunk 5]
```diff
@@ -5,33 +5,84 @@
...
+import Darwin
+#endif
+import class Foundation.PropertyListDecoder
+import class Foundation.ProcessInfo
+import class Foundation.FileManager
 
+let env = ProcessInfo.processInfo.environment
+
+// This struct is read from the .plist stored at versionDependencies.plist
 struct VersionDeps: Codable {
```
[the remark this answers is not in the document]
Comment (jbelkins) [reply scripts/generatePackageSwift.swift]: Name change made, along with a couple other related renames in this script.

**Pull request merged**